### PR TITLE
Input

### PR DESCRIPTION
**Pull request by tmshl**

### CODE REVIEW

**Comment by tmshl** [comment]
Base function completed raspberries can now send data to server.
Please review and merge if ok.

File: application/controller/ControllerRaspi.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+		$this->redirect(new \MVC\Route("raspi", "userinput"));
+	}
+      
+   /**
+    * Allows computers to automatically post error messages
+    */
+   function actionInput(\MVC\RequestHttp $request) {
+		$response = new \MVC\ResponseHtml($request->route, "main");
+      
+      $dbConfig = \MVC\MVC::getInstance()->data->db;
```
Comment (douira) [inline application/controller/ControllerRaspi.php]: do we actually ever get instances of the datbases without giving them the default database config?

File: application/controller/ControllerRaspi.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+      $json->deviceStatus->overwrite("memory_total", (int)$json->deviceStatus->memory_total);
+      
+      $modules = ModuleDb::getInstance($dbConfig)->getNameArray();      
+      foreach ($json->messages as $moduleName=>$messages) {
+         // check, if module names in JSON are compatible with the ones in the database
+         if (!in_array($moduleName, $modules)) {
+            throw new \MVC\exception\InputInvalid("Die gesendeten Daten liegen im falschen Format vor.");
+         }
+         
+         foreach($messages as $nr=>$message) {
```
Comment (douira) [inline application/controller/ControllerRaspi.php]: what is $nr used for here?

File: application/controller/ControllerRaspi.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+      $messages = array();
+      foreach ($json->messages as $moduleName => $rawMessages) {
+         $messageDbFactory->setDefaultModule($moduleDb->getByName($moduleName));
+         foreach ($rawMessages as $messageDataRow) {
+            $messages[] = $messageDbFactory->create($messageDataRow->toArray());
+         }
+      }
+      
+      // Save data to database
+      DeviceStatusDb::getInstance($dbConfig)->update($deviceStatus, $currDevice);
```
Comment (douira) [inline application/controller/ControllerRaspi.php]: does DeviceStatusDb identify/update devices by $currDevice? Shouldn't hte order be swapped then? 
(identifier, new value)

File: application/controller/ControllerRaspi.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      $messageDbFactory = new MessageDbFactory($messageTypes);
+      $messageDbFactory->setDefaultBugStatus($bugStatusNew);
+      
+      $moduleDb = ModuleDb::getInstance($dbConfig);
+      
+      $messages = array();
+      foreach ($json->messages as $moduleName => $rawMessages) {
+         $messageDbFactory->setDefaultModule($moduleDb->getByName($moduleName));
+         foreach ($rawMessages as $messageDataRow) {
+            $messages[] = $messageDbFactory->create($messageDataRow->toArray());
```
Comment (douira) [inline application/controller/ControllerRaspi.php]: nope, nifty php feature ;-)

File: application/lib/DeviceDb.php [hunk 5]
```diff
@@ -100,4 +100,32 @@ public function getByIp(Ip $ip) : Device {
          return null;
       }
    }
+   
+   public function getByType() {
```
Comment (douira) [inline application/lib/DeviceDb.php]: is this supposed to select by type of message?

Comment (douira) [inline application/lib/DeviceDb.php]: we're going to change this!

File: application/controller/ControllerRaspi.php (hunk 3 above)
Comment (douira) [inline application/controller/ControllerRaspi.php]: we'll change this.

**Pull request merged**